### PR TITLE
[AIRFLOW-3813] Add CLI commands to manage roles

### PR DESCRIPTION
**Pull request by astahlman**
Make sure you have checked _all_ steps below.

### Jira

- [ X ] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-3813
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.

### Description

- [ X ] Here are some details about my PR, including screenshots of any UI changes:

Here is the help text of the new command `airflow roles`:

    usage: airflow roles [-h] [-c] [-d] [-l] [role [role ...]]

    positional arguments:
      role          The name of a role

    optional arguments:
      -h, --help    show this help message and exit
      -c, --create  Create a new role
      -d, --delete  Delete an existing role
      -l, --list    List roles

Create and delete are both reentrant, i.e., `create` only adds a new
role if it does not exist, and `delete` only removes roles that do
exist.

### Tests

- [ X ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

Added unit tests for creating, deleting, and listing roles in tests/core.py

### Commits

- [ X ] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ X ] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.
  - All the public functions and the classes in the PR contain docstrings that explain what it does

### Code Quality

- [ X ] Passes `flake8`

### CODE REVIEW

**Comment by feng-tao** [comment]
CI seems to fail

File: airflow/www/security.py [hunk 1]
```diff
@@ -195,6 +195,19 @@ def init_role(self, role_name, role_vms, role_perms):
         else:
             self.log.info('Existing permissions for the role:%s within the database will persist.', role_name)
 
+    def delete_role(self, role_name):
```
Comment (feng-tao) [inline airflow/www/security.py:198]: Have you checked if FAB's SecurityManager provided any API to delete role?

File: airflow/www/security.py [hunk 2]
```diff
@@ -195,6 +195,19 @@ def init_role(self, role_name, role_vms, role_perms):
...
+        session = self.get_session
+        role = session.query(sqla_models.Role)\
+                      .filter(sqla_models.Role.name == role_name)\
+                      .first()
+        if role:
+            self.log.info("Deleting role '{}'".format(role_name))
+            session.delete(role)
```
Comment (feng-tao) [inline airflow/www/security.py:209]: I am wondering what will happen if a user's role gets deleted. Will the user has no role associated in the db?

File: airflow/bin/cli.py [hunk 3]
```diff
@@ -2296,8 +2339,11 @@ class CLIFactory(object):
                      'add_role', 'remove_role', 'user_import', 'user_export',
                      'username', 'email', 'firstname', 'lastname', 'role',
                      'password', 'use_random_password'),
-        },
-        {
+        }, {
+            'func': roles,
+            'help': 'Create/Delete roles',
```
Comment (feng-tao) [inline airflow/bin/cli.py]: Create / Delete / List roles?

File: airflow/www/security.py (hunk 2 above)
Comment (feng-tao) [reply airflow/www/security.py:209]: There is a table called ab_user_role which is a join result between ab_user and ab_role. What will happen to that table in this case?

File: airflow/www/security.py (hunk 1 above)
Comment (feng-tao) [reply airflow/www/security.py:198]: And maybe we need to check and make sure that user won't delete those 5 default roles defined in the security.py.

**Review by feng-tao — COMMENTED** [review]
hey @astahlman , I put up some comments on the pr. Mostly around what happens when a role is deleted. If the issue is not easy to address, we could split the pr into create role and delete role. And we could work on the create role and then work on delete role later(maybe worth to check with FAB committer).

File: airflow/www/security.py (hunk 1 above)
Comment (astahlman) [reply airflow/www/security.py:198]: Yeah, it doesn't appear to. These are the methods it does have:

```
    """
    ----------------------
     PRIMITIVES FOR ROLES
    ----------------------
    """
    def find_role(self, name):
        raise NotImplementedError

    def add_role(self, name):
        raise NotImplementedError

    def get_all_roles(self):
        raise NotImplementedError
```

From https://flask-appbuilder.readthedocs.io/en/latest/_modules/flask_appbuilder/security/manager.html

File: airflow/www/security.py (hunk 2 above)
Comment (astahlman) [reply airflow/www/security.py:209]: Yeah, delete will cascade to the `ab_user_role` table, so it just deletes any entries associated with that role. Here's an example:

```
$ airflow users -l | grep "jdoe+1"
│    3 │ user1          │ jdoe+1@gmail.com │ jon          │ doe1        │ [Admin, Op, Public] │

$ airflow roles --delete Public
[2019-02-06 13:56:24,911] {__init__.py:51} INFO - Using executor SequentialExecutor
[2019-02-06 13:56:25,431] {__init__.py:298} INFO - Filling up the DagBag from /Users/andrewstahlman/airflow/dags
[2019-02-06 13:56:25,638] {security.py:452} INFO - Start syncing user roles.
[2019-02-06 13:56:25,684] {security.py:196} INFO - Existing permissions for the role:Viewer within the database will persist.
[2019-02-06 13:56:25,729] {security.py:196} INFO - Existing permissions for the role:User within the database will persist.
[2019-02-06 13:56:25,771] {security.py:196} INFO - Existing permissions for the role:Op within the database will persist.
[2019-02-06 13:56:25,772] {security.py:365} INFO - Fetching a set of all permission, view_menu from FAB meta-table
[2019-02-06 13:56:25,927] {security.py:316} INFO - Cleaning faulty perms
[2019-02-06 13:56:25,934] {security.py:204} INFO - Deleting role 'Public'

$ airflow users -l | grep "jdoe+1"
│    3 │ user1          │ jdoe+1@gmail.com │ jon          │ doe1        │ [Admin, Op] │
```

File: airflow/bin/cli.py (hunk 3 above)
Comment (astahlman) [reply airflow/bin/cli.py]: Good point, will add

**Comment by feng-tao** [comment]
LGTM, thanks @astahlman . Maybe create another ticket to keep track of the delete role given we only handle create / list role here? We could either handle the delete_role  in Airflow in a separate pr or submit a pr in FAB to support that API and use that api for delete role.

oh, could you also update https://github.com/apache/airflow/blob/master/docs/howto/add-new-role.rst given now we could use cli to add role as well?

**Comment by astahlman** [comment]
@feng-tao I've updated the docs and created https://issues.apache.org/jira/browse/AIRFLOW-3818 to track the `--delete` command

**Comment by feng-tao** [comment]
The CI error is unrelated.

**Comment by geez** [comment]
@feng-tao Hello, the screenshot posted on that page https://airflow.readthedocs.io/en/latest/howto/add-new-role.html, is that `1.10.2`?
I installed `1.10.2` to test RBAC and DAG level access but cannot see the UI described in the screenshot (security tabs, icons, etc...).

Cheers,
G.

**Pull request merged**